cli/dispatch_minimal.rs

```diff
@@ -0,0 +1,132 @@
+  s[0] == DISPATCH_MINIMAL_TOKEN
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+// This corresponds to RecordMinimal on the TS side.
+struct Record {
+  pub promise_id: i32,
+  pub op_id: i32,
+  pub arg: i32,
+  pub result: i32,
+}
+
+impl Into<Buf> for Record {
+  fn into(self) -> Buf {
+    let vec = vec![
+      DISPATCH_MINIMAL_TOKEN,
+      self.promise_id,
+      self.op_id,
+      self.arg,
+      self.result,
+    ];
+    let buf32 = vec.into_boxed_slice();
+    let ptr = Box::into_raw(buf32) as *mut [u8; 5 * 4];
+    unsafe { Box::from_raw(ptr) }
+  }
+}
+
+impl From<&[i32]> for Record {
+  fn from(s: &[i32]) -> Record {
+    let ptr = s.as_ptr();
+    let ints = unsafe { std::slice::from_raw_parts(ptr, 5) };
+    assert_eq!(ints[0], DISPATCH_MINIMAL_TOKEN);
+    Record {
+      promise_id: ints[1],
+      op_id: ints[2],
+      arg: ints[3],
+      result: ints[4],
+    }
+  }
+}
+
+pub fn dispatch_minimal(
+  state: &ThreadSafeState,
+  control32: &[i32],
+  zero_copy: Option<PinnedBuf>,
+) -> Op {
+  let record = Record::from(control32);
+  let is_sync = record.promise_id == 0;
+  let min_op = match record.op_id {
+    OP_READ => ops::read(record.arg, zero_copy),
+    OP_WRITE => ops::write(record.arg, zero_copy),
+    _ => unimplemented!(),
+  };
+
+  let state = state.clone();
+  let mut record_ = record;
+
+  let fut = Box::new(min_op.then(move |result| -> Result<Buf, ()> {
+    match result {
+      Ok(r) => {
+        record_.result = r;
+      }
+      Err(err) => {
+        // TODO(ry) The dispatch_minimal doesn't properly pipe errors back to
+        // the caller.
+        debug!("swallowed err {}", err);
+        record_.result = -1;
+      }
+    }
+    let buf: Buf = record_.into();
+    state.metrics_op_completed(buf.len());
+    Ok(buf)
+  }));
+  if is_sync {
+    Op::Sync(fut.wait().unwrap())
+  } else {
+    Op::Async(fut)
+  }
+}
+
+mod ops {
+  use crate::errors;
+  use crate::resources;
+  use crate::tokio_write;
+  use deno::PinnedBuf;
+  use futures::Future;
+
+  type MinimalOp = dyn Future<Item = i32, Error = errors::DenoError> + Send;
+
+  pub fn read(rid: i32, zero_copy: Option<PinnedBuf>) -> Box<MinimalOp> {
+    debug!("read rid={}", rid);
+    let zero_copy = zero_copy.unwrap();
+    match resources::lookup(rid as u32) {
+      None => Box::new(futures::future::err(errors::bad_resource())),
+      Some(resource) => Box::new(
+        tokio::io::read(resource, zero_copy)
+          .map_err(errors::DenoError::from)
+          .and_then(move |(_resource, _buf, nread)| Ok(nread as i32)),
+      ),
+    }
+  }
+
+  pub fn write(rid: i32, zero_copy: Option<PinnedBuf>) -> Box<MinimalOp> {
+    debug!("write rid={}", rid);
+    let zero_copy = zero_copy.unwrap();
+    match resources::lookup(rid as u32) {
+      None => Box::new(futures::future::err(errors::bad_resource())),
+      Some(resource) => Box::new(
+        tokio_write::write(resource, zero_copy)
+          .map_err(|err| err.into())
```

Comment:
This also gives me a 'redundant closure' warning. Use `.map_err(DenoError::from)`.

Reply:
done